pika/adapters/blocking_connection.py

```diff
@@ -354,40 +351,31 @@ def __init__(self, parameters=None, _impl_class=None):
         # `_request_channel_dispatch`
         self._channels_pending_dispatch = set()
 
-        # Receives on_open_callback args from Connection
-        self._opened_result = _CallbackResult(self._OnOpenedArgs)
-
-        # Receives on_open_error_callback args from Connection
-        self._open_error_result = _CallbackResult(self._OnOpenErrorArgs)
-
         # Receives on_close_callback args from Connection
         self._closed_result = _CallbackResult(self._OnClosedArgs)
 
-        # Set to True when when user calls close() on the connection
-        # NOTE: this is a workaround to detect socket error because
-        # on_close_callback passes reason_code=0 when called due to socket error
-        self._user_initiated_close = False
-
-        impl_class = _impl_class or SelectConnection
-        self._impl = impl_class(
-            parameters=parameters,
-            on_open_callback=self._opened_result.set_value_once,
-            on_open_error_callback=self._open_error_result.set_value_once,
-            on_close_callback=self._closed_result.set_value_once)
-
-        self._impl.ioloop.activate_poller()
-
-        self._process_io_for_connection_setup()
+        # Perform connection workflow
+        self._impl = None  # so that attribute is created in case below raises
+        self._impl = self._create_connection(parameters, _impl_class)
+        self._impl.add_on_close_callback(self._closed_result.set_value_once)
 
     def __repr__(self):
         return '<%s impl=%r>' % (self.__class__.__name__, self._impl)
 
+    def __enter__(self):
+        # Prepare `with` context
+        return self
+
+    def __exit__(self, exc_type, value, traceback):
+        # Close connection after `with` context
+        if self.is_open:
```

Comment:
Added the `self.is_open` check because `BlockingConnection.close()` now raises the "wrong state" exception if it's not open.